### PR TITLE
[Flow] Add pass to fuse encoding ops into dispatch regions after hoisting

### PR DESCRIPTION
**Pull request by Max191**
This PR is the follow up to https://github.com/iree-org/iree/pull/18063, implementing the fusion pass to move set_encoding ops into producer dispatch regions when when the producer op is a LinalgOp. If there is no fusable producer, then the set_encoding is wrapped in a new dispatch region.

This PR also implements the ReifyRankedShapedTypeOpInterface for DispatchRegionOp, which is needed to resolve tensor dim ops after moving dynamically shaped SetEncoding ops into producer dispatch regions. The implementation simply takes the dynamic dims from the result dynamic dims of the dispatch region op.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h]: hmm, do we need this?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/test/fuse_encoding_ops_into_dispatch_regions.mlir
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/test/fuse_encoding_ops_into_dispatch_regions.mlir]: This part is not need in the test. Can we remove them. My concerns comes from maintenance burden, so it is better to simplify the tests.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp]: nit: `target`, not `targets`

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td]: I think we can improve the summary slightly if we put `producer` to it. E.g., `Fuses encoding ops into producer dispatch regions, or ...`

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FuseEncodingOpsIntoDispatchRegions.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FuseEncodingOpsIntoDispatchRegions.cpp]: I think what you're looking for is `encodingOp.getSourceMutable()`. It avoids the magic number `0`.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.cpp]: Similar to what I left in https://github.com/iree-org/iree/pull/18063#discussion_r1700879173, we don't need the insertion guard IMO.

**Review by hanhanW — DISMISSED** [review]
Overall looks good to me, just few nits.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/test/fuse_encoding_ops_into_dispatch_regions.mlir
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/test/fuse_encoding_ops_into_dispatch_regions.mlir]: Nice improvement! It makes the test better.

**Review by hanhanW — COMMENTED** [review]
LGTM, I'll approve the PR after we land https://github.com/iree-org/iree/pull/18063

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp]: Why is this dropping the comments and please keep these `static`.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h]: What does `overrideProducerResult` do?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.h
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.h:107]: How would you handle something like 

```
%r = flow.dispatch.region
%1 = op1(%r)
%2 = op2(%r, %1)
%3 = op3(%2)
```
and you want to move `%2` in?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.h
Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.h:107]: The caller should be responsible for fusing `%1` in before fusing `%2`. If the function is called on `op2` in that situation it should fail. I might not have added the necessary checks for that yet, though. I'll make sure it fails appropriately when I come back to address comments.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h
Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h]: When the producer is in a dispatch region, the defining op of the `OpOperand` will be the dispatch region.  `overrideProducerResult` is a way to circumvent this by acting as a replacement for the `OpOperand` defining op's result (it will get the matching yielded value inside the dispatch).

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp
Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp]: I moved it to FormDispatchRegions.h so I could reuse it in`FuseEncodingOpsIntoDispatchRegions`.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h]: Sorry, not able to follow. Can you give an example.

Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h]: In a case like:
```
%r = dispatch_region {
  %producer = op()
  flow.return %producer
}
%encoding = set_encoding %r
```
We want to check if `set_encoding` is fusable with `op`, but the OpOperand of `set_encoding` is `%r`. In this case, `overrideProducerResult` will be passed as `%producer`, and then the function will treat it as if it looks like:
```
%producer = op()
%encoding = set_encoding %producer
```
And decide whether the 2 ops are fusable.

Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h]: I can add this example to the comments on the function

Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.h]: I removed the dependency on FormDispatchRegions, and just checked that producers are linalg ops, as we discussed offline.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.h
Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/RegionOpUtils.h:107]: Returns failure in this case now.

**Review by MaheshRavishankar — APPROVED** [review]
Thanks Max for breaking up the PR and making this structured well!

**Pull request merged**
